src/calendar_duration.rs

```diff
@@ -0,0 +1,379 @@
+    /// // Duration that will precisely count the number of seconds, including leap seconds.
+    /// CalendarDuration::new().seconds(23_456).millis(789).unwrap();
+    /// ```
+    pub const fn new() -> Self {
+        Self {
+            months: 0,
+            days: 0,
+            seconds: 0,
+            nanos: expect!(NonZeroU32::new(1), "can't fail, non-zero"),
+        }
+    }
+
+    /// Set the months component of this duration to `months`.
+    pub const fn months(mut self, months: u32) -> Self {
+        self.months = months;
+        self
+    }
+
+    /// Set the months component of this duration to `years * 12 + months`.
+    ///
+    ///  # Errors
+    ///
+    /// Returns `None` if the value of the months component would be out of range.
+    pub const fn years_and_months(mut self, years: u32, months: u32) -> Option<Self> {
+        self.months = try_opt!(try_opt!(years.checked_mul(12)).checked_add(months));
+        Some(self)
+    }
+
+    /// Set the days component of this duration to `days`.
+    pub const fn days(mut self, days: u32) -> Self {
+        self.days = days;
+        self
+    }
+
+    /// Set the days component of this duration to `weeks * 7 + days`.
+    ///
+    ///  # Errors
+    ///
+    /// Returns `None` if the value of the days component would be out of range.
+    pub const fn weeks_and_days(mut self, weeks: u32, days: u32) -> Option<Self> {
+        self.days = try_opt!(try_opt!(weeks.checked_mul(7)).checked_add(days));
+        Some(self)
+    }
+
+    /// Sets the accurate component of this duration to a value in minutes and seconds.
+    ///
+    /// The definition of a minute is not exactly 60 seconds; it can also be 61 or 59 in the
+    /// presence of leap seconds.
+    ///
+    /// Durations made with this method are the closest thing to working like leap seconds don't
+    /// exists.
+    ///
+    /// # Errors
+    ///
+    /// Returns `None` if `seconds` exceeds 60, of if the value of the minutes component
+    /// (`hours * 60 + minutes`) would be out of range.
+    pub const fn hms(mut self, hours: u32, minutes: u32, seconds: u32) -> Option<Self> {
+        if seconds > 60 {
+            return None;
+        }
+        let minutes = try_opt!(try_opt!(hours.checked_mul(60)).checked_add(minutes));
+        if minutes > (u32::MAX >> 6) || seconds >= (1 << 6) {
+            return None;
+        }
+        self.seconds = minutes << 6 | seconds;
+        // Set the `0b10` flag that we store as discriminant in `nanos`.
+        self.nanos = expect!(
+            NonZeroU32::new(self.nanos.get() | 0b10),
+            "can't fail because the `| 0b10` ensures the value is non-zero"
+        );
+        Some(self)
+    }
+
+    /// Sets the accurate component of this duration to a value in seconds.
+    ///
+    /// This duration will count an exact number of seconds, which includes the counting of leap
+    /// seconds.
+    pub const fn seconds(mut self, seconds: u32) -> Self {
```

Comment:
If I understand correctly, the object is the builder so CalendarDuration::new().months(5).months(7) == CalendarDuration::new().months(7).

That is reasonably intuitive for months, but the struct suggests fields months, days, minutes, seconds, nanos. Because of that suggestion, I don't think a user would fully suspect CalendarDuration::new().hms(1, 2, 3).seconds(0) == CalendarDuration::new().